docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,263 @@
+```
+
+```mermaid
+graph TD
+    B-->A
+    C-->A
+    D-->A
+```
+
+We could either define an edge between every service to that service or have a special list that contains allIncomingAllowed.
+Services from that list would always be added as an outgoing edge.
+
+- `Mesh/MeshService`
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-mesh-to-ms
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    A-->B
+    A-->C
+    A-->D
+```
+
+This is simply `A` can talk to anyone, this could be either adding and outgoing edge from this service to every other service or having it on a list called allOutgoing and not bothering to trim the configuration.
+
+3. `MeshService/MeshService` - already talked about, this is the easy case
+
+4. `MeshServiceSubset` 
+
+Unfortunately there is one option here,
+we need to treat `MeshServiceSubset` just like `MeshService`.
+This is because we can't trim individual endpoints because the data is not available at that point, and
+we have to accept the fact that we will miss out on (small) parts of the config that could be trimmed.
+
+5. `MeshSubset`
+
+This one is a bit more tricky because the potential radious is a bit bigger.
+The approach has to be the same as in `MeshServiceSubset`, meaning: treat `MeshSubset` as just `Mesh`.
+
+There is a case that makes sense using `k8s.kuma.io/namespace: xyz` to allow only traffic from a specific namespace,
+so this needs to be documented that it won't improve the performance.
+
+#### Migration
+
+A typical migration to Mesh involves bringing services gradually into the Mesh.
+If the system is small enough, after several iterations, it will be enabled for the whole system (every service).
+
+If the system is too big to do this we need to do extra work in the iterations (defining MeshTrafficPermissions along the way).
+
+1. 0 services are in the Mesh
+2. Bring X services into the Mesh (where X is small enough, so the Mesh can handle it)
+3. Enable MeshTrafficPermission for services brought into the mesh (the performance improves because we trimmed the configs)
+4. Repeat point 2-3 until all services are in the Mesh
+
+The good thing about how this migration is set up is that we can gradually do it without any interruption to the system.
+The downside of it is that **a service owner needs to know EXACTLY which services consume their service** to avoid any disruptions.
+The migration assumes using Kubernetes DNS.
+
+Let's take the following service graph as an example:
+
+```mermaid
+flowchart LR
+ingress --> a --> b & c 
+c --> d & e
+b --> f
+```
+
+We want to bring services into the mesh one by one and trim their config, let's start with service `c` (it's a good example because it has both inbound and outbound traffic):
+0. We have a default MTP `Mesh/Mesh` with `allow`
```

Comment:
I think that's the broken premise here. The MTP allow all should only be used in demos. Service owners should think about who can access their services (if you do the code parallel it's similar to public/private).

Reply:
It's sort of similar to my previous point. I think that we also had this assumption when talking with @jakubdyszkiewicz

Reply:
> The MTP allow all should only be used in demos

_Meanwhile in the real world .. ;)_

Reply:
removed it as a premise since I checked via E2E test is not needed